src/nvim/normal.c

```diff
@@ -3940,6 +3957,101 @@ void scroll_redraw(int up, long count)
   redraw_later(VALID);
 }
 
+///
+/// Return length of line "lnum" for horizontal scrolling.
+///
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+  char_u *p = ml_get(lnum);
+  colnr_T col = 0;
+  if (*p != NUL) {
+    for (;;) {
+      int w = chartabsize(p, col);
+      mb_ptr_adv(p);
+      if (*p == NUL) {    // don't count the last character
+        break;
+      }
+      col += w;
+    }
+  }
+  return col;
+}
+
+///
+/// Find longest visible line number.  If this is not possible (or not desired,
+/// by setting 'h' in "guioptions") then the current line number is returned.
+///
+static linenr_T gui_find_longest_lnum(void)
+{
+  linenr_T ret = 0;
+
+  // Calculate maximum for horizontal scrollbar.  Check for reasonable
+  // line numbers, topline and botline can be invalid when displaying is
+  // postponed.
+  if (curwin->w_topline <= curwin->w_cursor.lnum &&
+      curwin->w_botline > curwin->w_cursor.lnum &&
+      curwin->w_botline <= curbuf->b_ml.ml_line_count + 1) {
+    linenr_T    lnum;
+    long      max = 0;
+
+    // Use maximum of all visible lines.  Remember the lnum of the
+    // longest line, closest to the cursor line.  Used when scrolling
+    // below.
+    for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum) {
+      colnr_T n = scroll_line_len(lnum);
+      if (n > (colnr_T)max) {
+        max = n;
+        ret = lnum;
+      } else if (n == (colnr_T)max
+          && abs((int)(lnum - curwin->w_cursor.lnum))
+          < abs((int)(ret - curwin->w_cursor.lnum))) {
+        ret = lnum;
+      }
+    }
+  } else {
+    // Use cursor line only.
+    ret = curwin->w_cursor.lnum;
+  }
+
+  return ret;
+}
+
+static linenr_T longest_lnum;
+///
+/// Do a horizontal scroll.  Return TRUE if the cursor moved, FALSE otherwise.
+///
+int gui_do_horiz_scroll(int leftcol, bool compute_longest_lnum)
```

Comment:
This and the other 2 new functions can be moved to `mouse.c` after adding these includes to mouse.c:

    #include "nvim/memline.h"
    #include "nvim/charset.h"